src/host/construction.jl

```diff
@@ -58,3 +58,6 @@ end
 
 Base.one(x::AbstractGPUMatrix{T}) where {T} = _one(one(T), x)
 Base.oneunit(x::AbstractGPUMatrix{T}) where {T} = _one(oneunit(T), x)
+
+Base.iszero(x::AbstractGPUMatrix{T}) where {T} = all(iszero, x)
+Base.isone(x::AbstractGPUMatrix{T}) where {T} = iszero(x-I)
```

Comment:
I think this will fail for non-square matrices:
```
julia> iszero([1 0 0; 0 1 0])
false

julia> all(iszero, [1 0 0; 0 1 0] - I)
ERROR: DimensionMismatch: matrix is not square: dimensions are (2, 3)
Stacktrace:
 [1] checksquare
   @ ~/.julia/dev/julia/usr/share/julia/stdlib/v1.9/LinearAlgebra/src/LinearAlgebra.jl:237 [inlined]
 [2] +(A::Matrix{Int64}, J::UniformScaling{Int64})
   @ LinearAlgebra ~/.julia/dev/julia/usr/share/julia/stdlib/v1.9/LinearAlgebra/src/uniformscaling.jl:215
```
But also, ideally it would not allocate a copy? Or perhaps that's OK as a fallback definition?

Reply:
There is no identity matrix for rectangular matrices, so I did expect it to crash for non-square matrices. I took  a look at LinearAlgebra's `isone` and, unlike what I thought, the convention is to return false for non-square matrices, so we could just do:
```
function Base.isone(x::AbstractGPUMatrix{T}) where {T}
        n,m = size(x)
        m != n && return false
        all(iszero, x-I)
end
```
I do agree this should be a fallback definition as it allocates a copy, but I couldn't code a kernel to do it the proper way. I am quite new to GPU programming, and still haven't understood how to stop the entire kernel should one thread meet a requirement (in this case, a non-diagonal term is not 0, or a diagonal term is not 1).

Reply:
There's no real way to stop an entire kernel if one thread encounters a specific condition. You can only synchronize threads within a block, so other blocks might have completed already. There's solutions (like using a single-element flag array you check in the CPU), but I doubt those are valuable here.

Reply:
I'm a bit surprised, as I expected this type of operation to be easy to make in parrallel. If I understand correctly, it's the asynchronous part that makes it difficult, right? In this case, each block is independent, so there is no real way to stop a block if an other one has finished its part.

Reply:
I'm not sure what you mean by 'the asynchronous part'. Blocks are pretty much independent indeed, launched on separate processors in the GPU, so you don't want to have those depend on one another.

You can do this kind of operation using a reduction though, see https://github.com/JuliaGPU/GPUArrays.jl/pull/410. Maybe you can just reuse ==/isequal, relaxing it a bit so that you can pass `I` to it? Or duplicate the parts of that kernel you need here.

Reply:
There must be something I haven't understood, as I can't get it to work. The problem I see with a reduction is that `_mapreduce` expects its arguments to have a size, length, to be able to use `LinearIndices` and so on. This is not the case for LinearAlgebra's `I`, so I didn't manage to use a reduction to do what I want.